### ISSUE TITLE
Crash Unhandled exception windows

### ISSUE THREAD

**Issue by thirdmenson** [google/filament#673]
So I am developing a plugin a for a software and was trying to use filament api for graphics engine. I was able to build Suzanne using the filament static libs and include headers and Suzanne source code in visual studio 2017 and it runs fine in both release and debug builds. Then next step was to try to run suzanne code within the plugin code project but I am getting crashes so I am trying to debug it. 

I am getting access violation reading location 0xFFFFFFFFFFFFFF when I try to set intensity on the ibl in the suzanne example code
` auto ibl = FilamentApp::get().getIBL()->getIndirectLight();
        ibl->setIntensity(100000);`

The mIntensity shows as "unable to read memory"  when I break and try to debug it in the IndirectLight.h.  First I thought it might be having problem casting to FIndirectLight using upcast but then If I comment out that IBL code then upcasting tcm.getinstance() runs fine without error. 

I also had code to add two more Suzanne model with different default material like so: 
```
filament::MaterialInstance* mi = FilamentApp::get().getDefaultMaterial()->createInstance();
		mi->setParameter("baseColor", RgbaType::LINEAR, LinearColorA{ 1, 1, 1, 1 });
		mi->setParameter("roughness", 0.01f);
		mi->setParameter("metallic", 1.0f);
```
I get the same unhandled exception error when it tries to createInstance().

I have made sure to add sdl2.dll and the envs(ibl) folder that has the ibl in the host application main directory so I think it should be able to find the ibl. Not sure what's going on as suzanne debug build project without the host application seems to work fine. Any idea what's going on ? Thanks for you help !

**Comment by bejado** [comment]
What's the host application and how are you initializing the engine? It's tough to debug this without seeing more of your code. A stack trace would also be helpful.

**Comment by thirdmenson** [comment]
The host application is After effects. I am using the exact suzanne sample app code to make sure I can replicate things first as a plugin. So just like in FilamentApp like so : mFilamentApp->mEngine = Engine::create(config.backend);
![image] 

You can see the call stack below:  (https://user-images.githubusercontent.com/36080456/50621860-d3c1e280-0ed6-11e9-9422-6af60a69c032.png)

**Comment by bejado** [comment]
After Effects calls plugins from at least 2 different threads. Are you initializing filament on the same thread you’re using to call setIntensity? Filament methods can only be called from a single thread.

**Comment by thirdmenson** [comment]
I am calling filament code in DoClick() after effects PF_Event_DO_CLICK click event. I didn't realize doclick was being called by two separate thread. If that's the case then that's why the crash is happening.  Need to figure out how to call on the same thread in after effects.

**Comment by thirdmenson** [comment]
Why do you think ae calls plugins from 2 different threads? Is it because there is "external code" in callstack between those two calls? Here is the filament suzanne app with static lib without the host application which runs completely fine but also has that "external code" in between those two calls .
![image](https://user-images.githubusercontent.com/36080456/50623004-be9d8180-0edf-11e9-9a20-924577d209f9.png)

I also looked up the after effects sdk and it only has two threads UI and Render thread. So anything on effects ui like click event should only be within the UI thread.

**Comment by thirdmenson** [comment]
Actually I found out that the debug suzanne static lib build without host app also crashes on exit. Looks like the material gets destroyed before the instances according to comment. I remember the debug static build without host app once did crash exactly like the one in host app but after that it just crashes on exit now. There might be some problem with debug build.

![image](https://user-images.githubusercontent.com/36080456/50626172-b4d54780-0efa-11e9-8ad5-0c5ff95ff8ba.png)

**Comment by bejado** [comment]
> debug suzanne static lib build without host app

I'm not sure what you mean by this. Do you mean the unmodified Suzanne demo? I'm able to run that in both Debug and Release without any crashes. (There was a crash upon exiting (#674), but I don't think that's related.)

In order to help I need to see a minimum code sample that reproduce the issue.

**Comment by thirdmenson** [comment]
Yes it's slight modified Suzanne demo but it's using static libraries as opposed to the full source code.  Here is the link to the whole visual studio 2017 project that crashes on exit in debug build. 

https://drive.google.com/file/d/14zvrZ52bmT34tsOQsSRJShMpXT6jG1ZL/view?usp=sharing

I am not sure if this crash is related to the after effects plugin project because the ultimate goal is to make is suzanne demo work inside an after effects plugin.

**Comment by thirdmenson** [comment]
I am surprised but I am able to launch Suzanne demo code inside after effects with code below, but as soon as I add the commented portion of the code: setting ibl intensity or getting default material I crash with same error I mentioned in the first post.

```
Config config;
	config.title = "suzanne";
	// config.backend = Backend::VULKAN;
	config.iblDirectory = FilamentApp::getRootPath() + IBL_FOLDER;
	
	App app;
	auto setup = [config, &app](Engine* engine, View* view, Scene* scene) {
		auto& tcm = engine->getTransformManager();
		auto& rm = engine->getRenderableManager();
		auto& em = utils::EntityManager::get();

		// Create textures. The KTX bundles are freed by KtxUtility.
		auto albedo = new image::KtxBundle(TEXTURES_ALBEDO_S3TC_DATA, TEXTURES_ALBEDO_S3TC_SIZE);
		auto ao = new image::KtxBundle(TEXTURES_AO_DATA, TEXTURES_AO_SIZE);
		auto metallic = new image::KtxBundle(TEXTURES_METALLIC_DATA, TEXTURES_METALLIC_SIZE);
		auto normal = new image::KtxBundle(TEXTURES_NORMAL_DATA, TEXTURES_NORMAL_SIZE);
		auto roughness = new image::KtxBundle(TEXTURES_ROUGHNESS_DATA, TEXTURES_ROUGHNESS_SIZE);
		app.albedo = KtxUtility::createTexture(engine, albedo, true, false);
		app.ao = KtxUtility::createTexture(engine, ao, false, false);
		app.metallic = KtxUtility::createTexture(engine, metallic, false, false);
		app.normal = KtxUtility::createTexture(engine, normal, false, false);
		app.roughness = KtxUtility::createTexture(engine, roughness, false, false);
		TextureSampler sampler(TextureSampler::MinFilter::LINEAR_MIPMAP_LINEAR,
			TextureSampler::MagFilter::LINEAR);

		// Instantiate material.
		app.material = Material::Builder()
			.package(RESOURCES_TEXTUREDLIT_DATA, RESOURCES_TEXTUREDLIT_SIZE).build(*engine);
		app.materialInstance = app.material->createInstance();
		app.materialInstance->setParameter("albedo", app.albedo, sampler);
		app.materialInstance->setParameter("ao", app.ao, sampler);
		app.materialInstance->setParameter("metallic", app.metallic, sampler);
		app.materialInstance->setParameter("normal", app.normal, sampler);
		app.materialInstance->setParameter("roughness", app.roughness, sampler);

		view->setAntiAliasing(View::AntiAliasing::NONE);
		view->setSampleCount(32);
		view->setDepthPrepass(View::DepthPrepass::DISABLED);

		//auto ibl = FilamentApp::get().getIBL()->getIndirectLight();
		//ibl->setIntensity(100000);

		
		// Add geometry into the scene.
		app.mesh = MeshReader::loadMeshFromBuffer(engine, RESOURCES_SUZANNE_DATA, nullptr, nullptr,
			app.materialInstance);
		auto ti = tcm.getInstance(app.mesh.renderable);
		app.transform = mat4f{ mat3f(1), float3(0, 0, -4) } *tcm.getWorldTransform(ti);
		rm.setCastShadows(rm.getInstance(app.mesh.renderable), false);
		scene->addEntity(app.mesh.renderable);
		tcm.setTransform(ti, app.transform);

		/*
		
		filament::MaterialInstance* mi = FilamentApp::get().getDefaultMaterial()->createInstance();
		mi->setParameter("baseColor", RgbaType::LINEAR, LinearColorA{ 1, 1, 1, 1 });
		mi->setParameter("roughness", 0.01f);
		mi->setParameter("metallic", 1.0f);
		//mi->setParameter("emissive", RgbaType::LINEAR, LinearColorA{ 1, 1, 1, 10 });

		app.mesh2 = MeshReader::loadMeshFromBuffer(engine, RESOURCES_SUZANNE_DATA, nullptr, nullptr, mi);
		auto ti2 = tcm.getInstance(app.mesh2.renderable);
		app.transform2 = mat4f{ mat3f::rotate(-0.8, float3{ 0, 1, 0 }), float3(2.0, 0, -3) } *tcm.getWorldTransform(ti2);
		rm.setCastShadows(rm.getInstance(app.mesh2.renderable), false);
		scene->addEntity(app.mesh2.renderable);
		tcm.setTransform(ti2, app.transform2);

		filament::MaterialInstance* mi2 = FilamentApp::get().getDefaultMaterial()->createInstance();
		mi2->setParameter("baseColor", RgbaType::LINEAR, LinearColorA{ 0, 0, 0, 1 });
		mi2->setParameter("roughness", 0.01f);
		mi2->setParameter("metallic", 0.5f);
		mi2->setParameter("reflectance", 0.9f);


		app.mesh3 = MeshReader::loadMeshFromBuffer(engine, RESOURCES_SUZANNE_DATA, nullptr, nullptr, mi2);
		auto ti3 = tcm.getInstance(app.mesh3.renderable);
		app.transform3 = mat4f{ mat3f::rotate(0.8, float3{ 0, 1, 0 }), float3(-2.0, 0, -3) } *tcm.getWorldTransform(ti3);
		rm.setCastShadows(rm.getInstance(app.mesh3.renderable), false);
		scene->addEntity(app.mesh3.renderable);
		tcm.setTransform(ti3, app.transform3);
		*/
		

	};

	auto cleanup = [&app](Engine* engine, View*, Scene*) {
		Fence::waitAndDestroy(engine->createFence());
		engine->destroy(app.materialInstance);
		engine->destroy(app.mesh.renderable);
		engine->destroy(app.material);
		engine->destroy(app.albedo);
		engine->destroy(app.normal);
		engine->destroy(app.roughness);
		engine->destroy(app.metallic);
		engine->destroy(app.ao);
	};

	FilamentApp::get().run(config, setup, cleanup);
```

**Comment by bejado** [comment]
The crash at `ibl->getIndirectLight()` occurs because `FilamentApp::get().getIBL()` returns `NULL`. Change the code to:
```
auto ibl = FilamentApp::get().getIBL();
assert(ibl);             // <-- will crash here
auto indirectLight = ibl->getIndirectLight();
indirectLight->setIntensity(100000);
...
```

to see what I mean. This is because your app can't find the IBL. You need to check your working directory, IBL path, etc.

The error message upon exit,
> reason: destroying material "aiDefaultMat" but 2 instances still alive".

occurs because you've created 2 new material instances (`mi` and `mi2`) but did not destroy them before exiting. To avoid this error, destroy them just like `app.materialInstance` is destroyed: `engine->destroy(app.materialInstance)`.

**Comment by thirdmenson** [comment]
I just tried your code and the ibl pointer is not null so the assert does not fail. I also checked the ibl path : C:\Program Files\Adobe\Adobe After Effects CC 2015\Support Files\envs\venetian_crossroads and I have ibl ktx files in that path

I see I need to manually delete everything I create.  Thanks.

**Comment by romainguy** [comment]
Note that the message about undestroyed objects is more a warning. If you destroy the engine all resources will be cleaned up automatically. It's there to indicate a potential programming errros.

**Comment by thirdmenson** [comment]
In the example project I uploaded it actually crashes not just a warning but this is not a big problem.

Crash on setting ibl params like intensity and rotation and when creating default material instance worries me. I made sure  the ibl is in the right path.  I also added the generated  folder that has resources/ textures and material in it to the host application directory. I am not sure if the generated folder is needed as  I thought all suzanne resources would get pulled from suzanne-resources.lib and common-resources.  The suzanne textures are rendering fine its the two default lit materials that I created crashes when creating instance: `filament::MaterialInstance* mi = FilamentApp::get().getDefaultMaterial()->createInstance();`

**Comment by bejado** [comment]
The default materials are not loaded from disk, they're compiled directly into `filament.lib`.

At this point you're able to run your modified standalone Suzanne demo successfully without crashing, correct? The 2 crashes (IBL / default material instance creation) only occur when you try to run Filament inside of an After Effects plugin?

**Comment by thirdmenson** [comment]
That is correct.

**Comment by thirdmenson** [comment]
It does not crash while creating engine material but it crashes further down the line when its rendering and trying to get the rotation on the indirect light.  The indirect light is corrupt anyways which we know from trying to set intensity. So why does the other filamentapp default material that has the different params crashes?

![image](https://user-images.githubusercontent.com/36080456/50672935-e573ba80-0fa8-11e9-8588-77674b737d19.png)

**Comment by romainguy** [comment]
Could you try stepping in the IBL.cpp code (it’s part of the samples source code) when the app is initialized so we can understand why you’re getting an invalid IBL?

**Comment by thirdmenson** [comment]
As you can see filament is finding the ibl and skybox ktx files. 
![image](https://user-images.githubusercontent.com/36080456/50674814-9a14d880-0fb7-11e9-9161-2301fe38143d.png)

It also looks like its building the IBL data structure correctly (at least there are no null pointers). 
![image](https://user-images.githubusercontent.com/36080456/50675098-e5c88180-0fb9-11e9-9ac7-8191ff27b42b.png)
![image](https://user-images.githubusercontent.com/36080456/50675158-363fdf00-0fba-11e9-9248-0ddea1393e8f.png)

And that's the end of loading IBL
![image](https://user-images.githubusercontent.com/36080456/50675230-b7977180-0fba-11e9-9da1-a6fceb2dd4bf.png)

Here is the image where we are trying to upcast from IndirectLight to FIndirectLight while trying to set intensity. To the left is the standalone filament suzanne app and to the right is the after effects plugin. If you look at the upcast returned the left one has valid IndirectLight, mreflectionsMapHandle and mIrradianceMapHandle but the after effects one on the right has unable to read memory.
![image](https://user-images.githubusercontent.com/36080456/50675572-f4fcfe80-0fbc-11e9-988b-e839685898fa.png)

Stepping over after this step, left one steps fine but right one crashes
![image](https://user-images.githubusercontent.com/36080456/50675597-207fe900-0fbd-11e9-8770-048d4fad9d56.png)

That's why I initially thought there was some problem upcasting. Not sure what's going on.

**Comment by bejado** [comment]
The issue isn't with `upcast`, it's with a corrupt `filament::IndirectLight` pointer. Note how the `this` pointer (`0x3b24d...`) inside of `setIntensity` differs from `mIndirectLight` (`0x0000...`) inside `IBL::loadFromKtx`.

I'm confused about the stack trace in your last screenshot. It seems like `FilamentApp:run` is calling out to After Effects handlers? What did you change inside of `FilamentApp::run`?

**Comment by thirdmenson** [comment]
You are correct but it seems like the mIndirectLight pointer inside ibl is correct if you look at the one on the right of the image below but when we get indirect light from ibl `ibl->getIndirectLight()` which goes through the upcast we get an invalid pointer for IndirectLight. The non-zero numbers are in the pointer are front somehow. That's weird. 

I haven't changed any code in FilamentApp::run as I am building against a static filament.lib. As you can see on the left part which is the standalone suzanne build that there is also SDL_main() code being called between FilamentApp::run and setting intensity code. I feel like it's because of the settupCallback() in FilamentApp::run.

![image](https://user-images.githubusercontent.com/36080456/50716714-dc3e2880-1050-11e9-9b83-0fba53a3f7bf.png)

**Comment by thirdmenson** [comment]
Finally I figured it out! Looks like the sample-app.lib was corrupt, so instead I imported the sample app source code itself and removed the sample-app.lib and no crash on set intensity. Material was still crashing, so instead of getting default material from Filament app I created the material in the main code like this: 
```
 app.mi1  = Material::Builder()
			.package(RESOURCES_AIDEFAULTMAT_DATA, RESOURCES_AIDEFAULTMAT_SIZE)
			.build(*engine)->createInstance();
```

Thank you Ben and Romain so much for your help. I really appreciate it !
